Mathlib/GameTheory/Auction/Secondprice.lean

```diff
@@ -0,0 +1,271 @@
+- `valuation_is_dominant`: Bidding one's valuation is a dominant strategy.
+- `first_price_has_no_dominant_strategy`: There is no dominant strategy in a first price auction.
+
+## Helping Lemmas
+
+- `gt_wins`: If `i`'s bid is higher than all other bids, then `i` wins.
+- `exists_max`: There exists a participant whose bid matches the highest bid
+- `winner_take_max`: The winner's bid is the highest.
+- `b_winner`: The winner's bid is at least the second highest bid.
+- `utility_winner`: If `i` wins, utility is the valuation minus the second highest bid.
+- `utility_loser`: If `i` does not win, their utility is 0.
+- `b_winner_max`: The winner's bid is greater than or equal to all other bids.
+- `b_loser_max`: If `i` does not win, the highest bid excluding `i` matches the highest bid.
+- `utility_first_price_winner`: If `i` wins in a first price auction utility is valuation minus bid.
+- `utility_first_price_loser`: If `i` does not win in a first price auction, utility is 0.
+
+## Notations
+
+- `|b|`: Represents a bidding function.
+- `maxb(b)`: The highest bid in the function `b`.
+- `B i`: The maximal bid of all participants but `i`.
+
+## Implementation Notes
+
+The structure and functions assume the existence of multiple bidders to allow for meaningful
+auction dynamics. Definitions like `winner` and `maxb` make use of Lean's `Finset` and `Classical`
+logic to handle potential non-constructive cases effectively.
+## References
+
+* [T. Roughgarden, *Twenty lectures on Algorithmic Game Theory*][roughgarden2016]
+
+## Tags
+
+auction, game theory, economics, bidding, valuation
+-/
+
+open Classical
+
+/-!### Structure Definition -/
+
+/-- The Auction structure is set with components including: -/
+structure Auction where
+   /-- A set of participants.-/
+   I : Type*
+   /-- A `Fintype` instance. -/
+   hF : Fintype I
+   /-- An `Inhabited` instance. -/
+   hI : Nontrivial I
+   /-- A function mapping each participant to their valuation. -/
+   v : I → ℝ
+
+attribute [instance] Auction.hF Auction.hI
+
+namespace Auction
+
+variable {a : Auction} (b : a.I → ℝ)
+
+/-!### Helper Functions and Definitions -/
+
+/-- Computes the highest bid given a bidding function `b`. -/
+@[simp]
+def maxb : ℝ := Finset.sup' Finset.univ Finset.univ_nonempty b
+
+/-- There exists a participant `i` whose bid equals the highest bid. -/
+lemma exists_max : ∃ i : a.I, b i = a.maxb b := by
+   obtain ⟨i, _, h2⟩ := Finset.exists_mem_eq_sup' Finset.univ_nonempty b
+   exact ⟨i, symm h2⟩
+
+/-- winner: The participant with the highest bid. -/
+noncomputable def winner : a.I := Classical.choose (exists_max b)
+
+/-- The bid of the winner equals the highest bid. -/
+lemma winner_take_max : b (winner b) = maxb b := Classical.choose_spec (exists_max b)
+
+/-- Removing a participant `i` from all participants still leaves a non-empty set. -/
+lemma delete_i_nonempty (i : a.I) : Finset.Nonempty (Finset.erase Finset.univ i) := by
+  obtain ⟨j, hj⟩ := exists_ne i
```

Comment:
I think mathlib is missing
```
lemma Finset.Nontrivial.univ {α} [Fintype α] [Nontrivial α] :
    Finset.Nontrivial (Finset.univ : Finset α) :=
  let ⟨a, b, h⟩ := exists_pair_ne α; ⟨a, mem_univ _, b, mem_univ _, h⟩
```
as then the proof here is just 
`Finset.Nontrivial.erase_nonempty (.univ)`

Reply:
Beat me to it :) I was just about to suggest adding something like
```
lemma Finset.univ_nontrivial_iff {α : Type*} [Fintype α] :
    (Finset.univ : Finset α).Nontrivial ↔ Nontrivial α := by
  rw [Finset.Nontrivial, Finset.coe_univ, Set.nontrivial_univ_iff]

lemma Finset.univ_nontrivial {α : Type*} [Fintype α] [h : Nontrivial α] :
    (Finset.univ : Finset α).Nontrivial :=
  Finset.univ_nontrivial_iff.mpr h
```